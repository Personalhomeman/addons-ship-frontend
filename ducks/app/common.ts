import { actionTypeCreator } from '@/utils';
export const $ = actionTypeCreator('APP');
