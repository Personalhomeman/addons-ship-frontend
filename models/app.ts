export type App = {
  appSlug: string;
  plan: string;
  title: string;
  avatarUrl: string;
  projectType: ProjectType;
  androidErrors: string[];
  iosErrors: string[];
  colors?: {
    start?: string;
    end?: string;
  };
};

export type ProjectType =
  | 'xamarin'
  | 'ios'
  | 'osx'
  | 'macos'
  | 'android'
  | 'cordova'
  | 'ionic'
  | 'react-native'
  | 'flutter'
  | 'other';
