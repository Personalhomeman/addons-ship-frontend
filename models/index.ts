import { NextContext } from 'next';
import { Store } from 'redux';
import { Request } from 'express';

export type AppVersionProps = {
  appSlug: string;
  versionId: string;
};

export const AppVersionPageTabs = ['details', 'devices', 'test', 'activity'];

export type AppVersionPageQuery = AppVersionProps & {
  isPublic?: string;
  selectedTab?: typeof AppVersionPageTabs[number];
};

export type AppPageQuery = {
  appSlug: string;
};

export type AppVersionList = Array<{
  version: string;
  appVersions: AppVersion[];
}>;

export type AppVersion = {
  id: number;
  version: string;
  platform: string;
  buildNumber: number;
  buildSlug: string;
  lastUpdate: Date;
  description: string;
  whatsNew: string;
  minimumOs: string;
  minimumSdk: string;
  packageName: string;
  certificateExpires: Date;
  bundleId: string;
  size: number;
  supportedDeviceTypes: string[];
  distributionType: string;
  iconUrl: string;
  appName: string;
  promotionalText: string;
  keywords: string;
  reviewNotes: string;
  supportUrl: string;
  marketingUrl: string;
  scheme: string;
  configuration: string;
  publicInstallPageURL: string;
};

export interface PageContext extends NextContext {
  store: Store;
  isServer: boolean;
  req: Request;
}

export type Platform = 'ios' | 'android';
