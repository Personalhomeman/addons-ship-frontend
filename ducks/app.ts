import { Dispatch } from 'redux';
import { createAction, createReducer } from 'deox';

import { ShipAPIService } from '@/services/ship-api';
import { actionTypeCreator } from '@/utils';
import { App } from '@/models/app';
import { placeholderAppIcon } from '@/config';

const $ = actionTypeCreator('APP');

const _fetchApp = (appSlug: string) => async (
  dispatch: Dispatch,
  _getState: any,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(fetchApp.next());

  try {
    const app = await shipApi.getApp(appSlug);

    dispatch(fetchApp.complete(app));
  } catch (error) {
    console.error('fetchApp', error);
    dispatch(fetchApp.error(error));
  }
};

export const fetchApp = Object.assign(_fetchApp, {
  next: createAction($`GET_NEXT`),
  complete: createAction($`GET_COMPLETE`, resolve => (app: App | null) => resolve(app)),
  error: createAction($`GET_ERROR`, resolve => (error: Error) => resolve(error))
});

const defaultState: AppState = null;

export type AppState = App | null;
export default createReducer(defaultState, handleAction => [
  handleAction(fetchApp.complete, (_, { payload }) => {
    if (payload) {
      const avatarUrl = payload.avatarUrl || placeholderAppIcon;
      return { ...payload, avatarUrl } as any;
    }
    return null;
  })
]);
