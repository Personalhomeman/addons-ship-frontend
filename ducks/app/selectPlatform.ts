import { createAction } from 'deox';

import { Platform } from '@/models';

import { $ } from './common';

export const selectPlatform = createAction($`SELECT_PLATFORM`, resolve => (platform: Platform) => resolve(platform));
