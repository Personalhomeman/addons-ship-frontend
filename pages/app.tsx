import { Component, Fragment } from 'react';
import { Dispatch, Store } from 'redux';
import { connect } from 'react-redux';

import { NextContext } from 'next';
import { NextDocumentContext } from 'next/document';

import { AppPageQuery } from '@/models';
import AppSummary from '@/components/AppSummary';
import SectionHeading from '@/components/SectionHeading';

interface AppPageProps extends AppPageQuery {}

type AppPageState = {};

interface Context extends NextDocumentContext {
  store: Store;
  isServer: boolean;
}

class AppPage extends Component<AppPageProps, AppPageState> {
  state: AppPageState = {};

  static getInitialProps({ query: { appSlug } }: NextContext) {
    return { appSlug };
  }

  render() {
    const { appSlug } = this.props;

    return (
      <Fragment>
        <AppSummary
          detailsPagePath="/v1-1-0/details"
          title="My app v1.1.0 (28)"
          description="This is a short description about my app. It is a good app. End of description."
          note="Updated on January 29, 2008"
          iconUrl="/static/icon-flutter.svg"
          platformIconUrl="/static/icon-apple.svg"
        />
        <SectionHeading>Version History</SectionHeading>
      </Fragment>
    );
  }
}

export default AppPage;
