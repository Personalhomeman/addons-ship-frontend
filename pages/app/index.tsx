import { Component } from 'react';
import { connect } from 'react-redux';

import { AppPageQuery, PageContext, AppVersionList } from '@/models';
import { RootState } from '@/store';
import { fetchAppVersionList } from '@/ducks/appVersionList';
import { getAppVersionsByVersion, getAppVersionsByBuildNumber } from '@/ducks/selectors';

import View from './view';

export interface AppPageProps extends AppPageQuery {
  appVersionsByVersion: Array<{
    groupName: string;
    appVersions: AppVersionList;
  }>;
  appVersionsByBuildNumber: Array<{
    groupName: string;
    appVersions: AppVersionList;
  }>;
}

type AppPageState = {
  selectedVersionSortingOptionValue: string | null;
};

export class AppPage extends Component<AppPageProps, AppPageState> {
  state: AppPageState = {
    selectedVersionSortingOptionValue: null
  };

  versionSortingOptions = [
    {
      text: 'Latest Build',
      value: 'latest-build'
    },
    {
      text: 'Latest version',
      value: 'latest-version'
    }
  ];

  componentDidMount() {
    this.setState({
      selectedVersionSortingOptionValue: 'latest-build'
    });
  }

  static async getInitialProps({ query: { appSlug }, store }: PageContext) {
    await store.dispatch(fetchAppVersionList(appSlug as string) as any);

    return { appSlug };
  }

  versionSortOptionWithValueSelected = (value: string) => {
    this.setState({
      selectedVersionSortingOptionValue: value
    });
  };

  render() {
    const groupedAppVersionList =
      this.state.selectedVersionSortingOptionValue && this.state.selectedVersionSortingOptionValue === 'latest-version'
        ? this.props.appVersionsByVersion
        : this.props.appVersionsByBuildNumber;
    const latestAppVersion = groupedAppVersionList[0].appVersions[0];
    const emptyPage = false;

    const viewProps = {
      emptyPage,
      latestAppVersion,
      versionSortingOptions: this.versionSortingOptions,
      versionSortOptionWithValueSelected: this.versionSortOptionWithValueSelected,
      selectedVersionSortingOption: this.versionSortingOptions.find(
        (versionSortingOption: any) => versionSortingOption.value === this.state.selectedVersionSortingOptionValue
      ),
      groupedAppVersionList
    };

    return <View {...viewProps} />;
  }
}

const mapStateToProps = (rootState: RootState) => ({
  appVersionsByVersion: getAppVersionsByVersion(rootState),
  appVersionsByBuildNumber: getAppVersionsByBuildNumber(rootState)
});

export default connect(mapStateToProps)(AppPage as any);
