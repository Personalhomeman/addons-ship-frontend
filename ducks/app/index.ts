import { Dispatch } from 'redux';
import { createAction, createReducer } from 'deox';

import { ShipAPIService } from '@/services/ship-api';
import { App } from '@/models/app';
import { placeholderAppIcon } from '@/config';
import { Platform } from '@/models';

import { $ } from './common';
import { selectPlatform } from './selectPlatform';

const _fetchApp = (appSlug: string) => async (
  dispatch: Dispatch,
  _getState: any,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(fetchApp.next());

  try {
    const app = await shipApi.getApp(appSlug);

    dispatch(fetchApp.complete(app));
  } catch (error) {
    console.error('fetchApp', error);
    dispatch(fetchApp.error(error));
  }
};

export const fetchApp = Object.assign(_fetchApp, {
  next: createAction($`GET_NEXT`),
  complete: createAction($`GET_COMPLETE`, resolve => (app: App | null) => resolve(app)),
  error: createAction($`GET_ERROR`, resolve => (error: Error) => resolve(error))
});

export { selectPlatform };

const defaultState: AppState = { app: null };

export type AppState = { app: App | null; selectedPlatform?: Platform };
export default createReducer(defaultState, handleAction => [
  handleAction(fetchApp.complete, (state, { payload }) => {
    if (payload) {
      const avatarUrl = `${payload.avatarUrl || placeholderAppIcon}?${Math.random()}`;

      const { androidErrors, iosErrors } = payload;
      return {
        ...state,
        selectedPlatform: 'ios',
        app: {
          ...payload,
          androidErrors: androidErrors || [],
          iosErrors: iosErrors || [],
          avatarUrl
        }
      } as any;
    }

    return null;
  }),
  handleAction(selectPlatform, (state, { payload }) => ({ ...state, selectedPlatform: payload }))
]);
