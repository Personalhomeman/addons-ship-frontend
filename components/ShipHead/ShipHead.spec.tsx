import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { App } from '@/models/app';

import { ShipHead } from '.';

describe('ShipHead', () => {
  test('When app is loaded', () => {
    const tree = shallowToJson(shallow(<ShipHead app={{ title: 'An app title' } as App}>Subpage</ShipHead>));
    expect(tree).toMatchSnapshot();
  });

  test('When app is not present', () => {
    const tree = shallowToJson(shallow(<ShipHead app={null}>Subpage</ShipHead>));
    expect(tree).toMatchSnapshot();
  });
});
