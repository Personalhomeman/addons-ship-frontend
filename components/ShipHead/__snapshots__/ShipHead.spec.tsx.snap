// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShipHead When app is loaded 1`] = `
<Head>
  <title>
    Subpage
     | An app title
  </title>
</Head>
`;

exports[`ShipHead When app is not present 1`] = `
<Head>
  <title>
    Subpage
  </title>
</Head>
`;
